### PR TITLE
Support TiXml 2.6.x

### PR DESCRIPTION
**Pull request by allenh1**
I got some rather disgusting compiler errors that come from
the old style of TiXml --> std::string process. They use the
visitor pattern now.

I'm not sure if this is compatible with the Ubuntu version
of TiXml, but, if not, I'll have to figure out a better way
to fix it.

### CODE REVIEW

File: urdf/include/urdf/model.h [hunk 1]
```diff
@@ -41,7 +41,7 @@
 #include <map>
 #include <memory>
 
-#include "tinyxml.h"  //NOLINT
+#include <tinyxml.h>
```
Comment (mikaelarguedas) [inline urdf/include/urdf/model.h:44]: I think this won't pass the linters, hence the original comment

File: urdf/src/model.cpp [hunk 2]
```diff
@@ -79,10 +79,9 @@ bool Model::initXml(TiXmlDocument * xml_doc)
     return false;
   }
 
-  std::stringstream ss;
-  ss << *xml_doc;
-
-  return Model::initString(ss.str());
+  TiXmlPrinter p;
+  return p.VisitEnter(std::cref(*xml_doc).get())
+    && Model::initString(std::string(p.CStr()));
```
Comment (mikaelarguedas) [inline urdf/src/model.cpp:84]: this won't pass linters

File: urdf/src/model.cpp [hunk 3]
```diff
@@ -92,10 +91,9 @@ bool Model::initXml(TiXmlElement * robot_xml)
     return false;
   }
 
-  std::stringstream ss;
-  ss << (*robot_xml);
-
-  return Model::initString(ss.str());
+  TiXmlPrinter p;  
+  return p.VisitEnter(std::cref(*robot_xml).get(), robot_xml->FirstAttribute())
+    && Model::initString(std::string(p.CStr()));
```
Comment (mikaelarguedas) [inline urdf/src/model.cpp:96]: this won't pass linters

**Review by mikaelarguedas — CHANGES_REQUESTED** [review]
Great thanks for reporting it! Unfortunately that won't be able to get in as is:
- these changes won't pass the ROS2 linters.
- They should be submitted upstream (the ROS1 repo) because they are not ROS2 specific.
- Ubuntu ships with tinyxml2 v4.0.1 since yakkety and the code compiles just fine (same for debian stretch), is this only a problem with 2.6? Or is it due to gcc8 ?

Please submit non ROS2 related changes to the ROS1 repository. If submitting ROS2 specific change, please first see the [contributing page](https://github.com/ros2/ros2/wiki/Contributing), the [developer guide](https://github.com/ros2/ros2/wiki/Developer-Guide) (including the testing section) and the [ROS2 onboarding document](https://github.com/ros2/ros2/wiki/ROS-2-On-boarding-Guide).
Finally please open your PR from a branch on this repo so that our CI can test the changes easily.

Thanks!

File: urdf/include/urdf/model.h (hunk 1 above)
Comment (allenh1) [reply urdf/include/urdf/model.h:44]: Ok, I thought that was related to the " instead of the '>'. I'll put the `NOLINT` comment back in place.

**Comment by allenh1** [comment]
> Ubuntu ships with tinyxml2 v4.0.1 since yakkety and the code compiles just fine (same for debian stretch), is this only a problem with 2.6? 

So this is using tinyxml, not tinyxml2. Yakkety uses tinyxml 2.6.2 currently. I'm very confused now. Do both libraries have the same include?

> Finally please open your PR from a branch on this repo so that our CI can test the changes easily.

I was going to, but I don't have push access to this one (I also don't have push access to the ROS2 fork of rosserial -- do you think this is because they are forked?).

**Comment by mikaelarguedas** [comment]
> So this is using tinyxml, not tinyxml2

Oh good point I was too fast on that one. 

> Yakkety uses tinyxml 2.6.2 currently. I'm very confused now. Do both libraries have the same include?

```
 libtinyxml-dev | 2.6.2-1build1 | precise/universe
 libtinyxml-dev | 2.6.2-2       | trusty/universe
 libtinyxml-dev | 2.6.2-2       | vivid
 libtinyxml-dev | 2.6.2-3       | xenial
 libtinyxml-dev | 2.6.2-3       | yakkety
 libtinyxml-dev | 2.6.2-4       | zesty
 libtinyxml-dev | 2.6.2-4       | artful

```
Looks like there have been no change in tinyxml on ubuntu for a while. given that we dont have any warning on xenial with this code I'd guess the problem is somewhere else.

> Do both libraries have the same include?

I'm fine including them as  system includes, but the need to respect google styleguide so that our jobs stay green.

I still believe the if the code change is necessary, it should go on the upstream repository and not this one.

**Comment by mikaelarguedas** [comment]
> I was going to, but I don't have push access to this one (I also don't have push access to the ROS2 fork of rosserial -- do you think this is because they are forked?).

How weird, I checked the permission and github says you have write access to this repo. For rosserial it's my bad, I didnt update the permissions so it stayed a read-only, now fixed

**Comment by mikaelarguedas** [comment]
After updating tinyxml on gentoo looks like this is not an issue anymore. Will close this for now and if change needs to happen it'll be done upstream and cherry-picked here

**Pull request closed without merging**